server/rest/src/main/java/org/infinispan/rest/stream/CacheChunkedStream.java

```diff
@@ -0,0 +1,205 @@
+         this.ctx = Objects.requireNonNull(ctx);
+         this.allocator = Objects.requireNonNull(allocator);
+      }
+
+      protected ByteBuf newByteBuf(int size) {
+         // Always need array to be 1 bigger than actual size for the ending bracket
+         int allocatedSize = Math.max(size + 1, CHUNK_SIZE);
+         return allocator.buffer(allocatedSize, allocatedSize);
+      }
+
+      @Override
+      protected void onStart() {
+         ByteBuf buf = newByteBuf(0);
+         buf.writeByte('[');
+         pendingBuffer.set(buf);
+         request(1);
+      }
+
+      @Override
+      public void onNext(T t) {
+         ByteBuf pendingBuf = pendingBuffer.get();
+         if (!firstEntry) {
+            pendingBuf.writeByte(',');
+         } else {
+            firstEntry = false;
+         }
+         ByteBuf newBuf = writeItem(t, pendingBuf);
+         if (newBuf != null) {
+            assert pendingBuf.writableBytes() == 0;
+            // It is technically possible that a single item may be larger than remaining pending and the new buffer
+            // In that case the implementation must return a larger ByteBuf that we provided, however it must also
+            // have at least a single writeable byte in case this is the last value and we need to append ] to it
+            assert newBuf.writableBytes() > 0;
+            pendingBuffer.set(newBuf);
+            readyBuffer.set(pendingBuf);
+
+            notifyReadyBuffer();
+         } else {
+            assert pendingBuf.writableBytes() > 0;
+            request(1);
+         }
+      }
+
+      // Implementation must write the bytes for the item to the pending buffer first. If the pending buffer does not
+      // contain enough bytes, the newByteBuf method should be invoked with how many bytes it requires and then write
+      // the same number of bytes into the returned ByteBuf and return the ByteBuf from this method
+      abstract ByteBuf writeItem(T item, ByteBuf pending);
+
+      @Override
+      public void onError(Throwable t) {
+         logger.error("Error encountered while streaming cache chunk", t);
+         // TODO: is there a way to propagate?
+      }
+
+      @Override
+      public void onComplete() {
+         ByteBuf buf = pendingBuffer.getAndSet(null);
+         buf.writeByte(']');
+         readyBuffer.set(buf);
+         completed = true;
+
+         notifyReadyBuffer();
+      }
+
+      private void notifyReadyBuffer() {
+         if (hasReturnedNull) {
+            hasReturnedNull = false;
+            ChunkedWriteHandler handler = ctx.pipeline().get(ChunkedWriteHandler.class);
+            if (handler != null) {
+               handler.resumeTransfer();
+            } else if (logger.isTraceEnabled()) {
+               logger.trace("Channel handler ChunkedWriteHandler not found to resume after stutter");
+            }
+         }
+      }
+
+      public void cancelSubscriber() {
+         cancel();
+         closeBuf(pendingBuffer);
```

Comment:
I guess the buffer could be released while `writeItem` executes? Not sure how much of a problem though.

Reply:
It would propagate as an error to onError below. Due to the onComplete above I am thinking of redoing how notifications would work, so I can cover this finally as well.